Unicorn stream8 feature bjs2 52697 correct. UserService - User Deactivation.

alobraztsov86 opened the pull request:
Следующие методы не реализованы, тк являются методами из других задач
```
EventService.deleteEvent()
GoalService.updateGoal()
GoalService.deleteGoal()
```

Commit:
BJS-52697 Deactivate user. Basic functionality of deactivation

Commit:
BJS-52697 Deactivate user.Autotests for UserController and UserService

Commit:
BJS-52697 Deactivate user.MentorshipService updated to use self written MentorshipRepository.deleteByMentorId() method

TFdevQA commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -0,0 +1,44 @@
+    @PostMapping("/deactivate")
+    @ResponseBody
+    public void deactivateUser(@RequestParam Long userId) {
+        if (idIsValid(userId)) {
```
Можно валидацию сделать через аннотацию (в user_service она уже добавлена)

alobraztsov86 replied:
Спасибо, попробовал с Min(1) потестил, но там еще по-хорошему обработу исключений нужно делать , поэтому решил
подобные задачи делать в рамках другой задачи, которая у меня как раз будет на валидаторы
https://faang-school.atlassian.net/browse/BJS2-52917

TFdevQA commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -0,0 +1,44 @@
+    @ResponseBody
+    public void deactivateUser(@RequestParam Long userId) {
+        if (idIsValid(userId)) {
+            userMapper.toDto(new User());
```
А зачем этот новый объект? Тем более нигде не исп-ся

alobraztsov86 replied:
Не удалил. Удалю.

TFdevQA commented on src/main/java/school/faang/user_service/mapper/UserMapper.java:
```diff
@@ -1,11 +1,14 @@
 
     UserDto toDto(User User);
+
+    User toEntity(UserDto userDto);
```
Ни одного места использования маппинга не нашла

alobraztsov86 replied:
Да, уже нету, уберу

TFdevQA commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -0,0 +1,44 @@
+@RequestMapping("/api/user")
+@Slf4j
+public class UserController {
+    @Autowired
```
По-моему излишне везде использовать @Autowired, бины же автоматоматом создадутся, если стоит на классе @Controller? И я бы использовала @RestController

alobraztsov86 replied:
Да, @Autowired лишним оказался, если есть конструктор дефолтный, @RestController - может тоже влияет.
Убрал @Autowired

Nafanuil replied:
вообще это к вопросу о том как проводить инъекцию бинов, через поля считается делать это не правильно и много статей на эту тему уже написано. вот к примеру одна маленькая из них https://habr.com/ru/articles/334636/

TFdevQA commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,112 @@
+
+    public void deactivateUser(Long userId) {
+        log.info(LOG_MESSAGE_DEACTIVATING_STARTS, userId);
+        User userToDeactivate = getUserFromDataBase(userId);
```
А если нет такого юзера в базе?

alobraztsov86 replied:
Если пользователя с таким userId нету, то метод getUserFromDataBase(userId) выбросит исключение

и на этом деактвация закончится, т.е. до запуска методов
quitGoals()
quitEvents()
quitMentorship()
выполнение не дойдёт - проверял.

TFdevQA replied:
Да, не увидела, что это твой метод, а не реквест в базу

Nafanuil replied:
> Если пользователя с таким userId нету, то метод getUserFromDataBase(userId) выбросит исключение
> 
> и на этом деактвация закончится, т.е. до запуска методов quitGoals() quitEvents() quitMentorship() выполнение не дойдёт - проверял.

есть один лайвхак как проверить свой код на душок, который говорит что изменение порядка строк в методе не должно менять его поведение (в книжках Чистый код и Рефакторинг)

реализовав это ты сможешь наглядно показать то что описал тут

alobraztsov86 replied:
Пока не понимаю, что нужно исправить и в чём выражается зависимость поведения от порядка строк метода, и какого метода.

TFdevQA commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,112 @@
+        log.info(LOG_MESSAGE_DEACTIVATING_STARTS, userId);
+        User userToDeactivate = getUserFromDataBase(userId);
+
+        quitGoals(userToDeactivate.getId());
```
Здесь же все тот же userId, зачем `userToDeactivate.getId()`?
И в следующем тоже самое

alobraztsov86 replied:
Если пользователя с таким userId нету, то метод getUserFromDataBase(userId) выбросит исключение

и на этом деактвация закончится, т.е. до запуска методов
quitGoals()
quitEvents()
quitMentorship()
выполнение не дойдёт - проверял.

alobraztsov86 replied:
Согдасен, так лучше. Поменяю

TFdevQA commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    private void quitGoals(Long userId) {
+        List<Goal> allUserGoals = goalRepository.findGoalsListByUserId(userId);
```
Опять нет проверки на наличие в базе

alobraztsov86 replied:
Если пользователя с таким userId нету, то метод getUserFromDataBase(userId) выбросит исключение
и на этом деактвация закончится, т.е. до запуска методов quitGoals() quitEvents() quitMentorship() выполнение не дойдёт - проверял.

TFdevQA commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,112 @@
+
+    private List<Goal> findGoalsToDelete(Long userId) {
+
+        List<Goal> userGoals = goalRepository.findGoalsListByUserId(userId);
```
ты это уже сделал же, зачем лишний раз ходить в базу за теммже самым?

alobraztsov86 replied:
Да, согласен, убираю. Но База может заскучать)))

TFdevQA commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,112 @@
+        List<Goal> goalsToDelete = findGoalsToDelete(userId);
+
+        for (Goal goal : goalsToDelete) {
+            goalService.deleteGoal(goal.getId());
```
А ты удаляешь, не проверив, что её не выполняют другие пользователи?

alobraztsov86 replied:
Проверяю в методе findGoalsToDelete()
Там если один пользователь у цели(т.е. с удаляемым userId), значит других пользователей у цели нет, только тогда добавляю в споисок удаляемыз целей

Nafanuil replied:
что то я не вижу там такого

alobraztsov86 replied:
В метод findGoalsToDelete() передаётся список целей деактивируемого пользователя.
т.е. у каждой цели метод getUsers() вернёт как минимум 1 пользователя - того которого деактивируем.
Поэтому, если у цели один пользователь - то это деактивируемый пользователь и других пользователей у цели нет. Поэтому я сократил преверку до if (users.size() == 1) в методе findGoalsToDelete()


изначально условие было
**if ((users.size() == 1) && (users.get(0).getId().equals(userId)) )**

До "оптимизации" код был такой(равнозначный текущему):
```
 private List<Goal> findGoalsToDelete(Long userId, List<Goal> userGoals) {
        List<Goal> goalsToDelete = new ArrayList<>();
        for (Goal goal : userGoals) {
            List<User> users = goal.getUsers();
            if ((users.size() == 1) && (users.get(0).getId().equals(userId)) ) {
                goalsToDelete.add(goal);
            }
        }
        return goalsToDelete;
    }
```
    
Верну его, так однозначнее читается условие

TFdevQA commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    private void quitEvents(Long userId) {
+        List<Event> eventsToDelete = eventRepository.findAllByUserId(userId);
```
Снова проверки нет

alobraztsov86 replied:
Если пользователя с таким userId нету, то метод getUserFromDataBase(userId) выбросит исключение
и на этом деактвация закончится, т.е. до запуска методов quitGoals() quitEvents() quitMentorship() выполнение не дойдёт - проверял.

alobraztsov86 replied:
Тут удаляю события без доп.проверок, тк так в задаче написано "Все события, запланированные пользователем должны быть остановлены и удалены." Остановить их неизвестно как, а удалить можно.

Nafanuil replied:
я бы тогда так и назвал метод stopAndDeleteEvent

alobraztsov86 replied:
Переименовал на stopAndDeleteEvent()

TFdevQA commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,112 @@
+        }
+    }
+
+    private void quitMentorship(User userToDeactivate) {
```
Сюда тоже можно передавать userId

alobraztsov86 replied:
Поменял

TFdevQA commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,112 @@
+
+        quitGoals(userToDeactivate.getId());
+        quitEvents(userToDeactivate.getId());
+        quitMentorship(userToDeactivate);
```
Зачем передавать User, чтобы вытаскивать id там, хотя id уже сразу есть

TFdevQA commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,112 @@
+        List<Goal> userGoals = goalRepository.findGoalsListByUserId(userId);
+        List<Goal> goalsToDelete = new ArrayList<>();
+        for (Goal goal : userGoals) {
+            List<User> users = goalRepository.findUsersByGoalIdHql(goal.getId());
```
А зачем из базы? Разве в сущности нет этого списка?

alobraztsov86 replied:
Есть вроде, поменяю. Спасибо.

Commit:
BJS2-52697 Deactivate user. Review changes. Autotests updated

Commit:
BJS2-52697 Deactivate user. Review changes. Autotests updated + MentorshipService.stopMentorship() updated; GoalRepository updated;

TFdevQA commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,109 @@
+@Transactional
+@Slf4j
+public class UserService {
+    @Autowired
```
Autowired в сервисе тоже можно не писать

Nafanuil replied:
смотри комментарий к инъекции в контроллере

alobraztsov86 replied:
Удалил Autowired

Nafanuil commented on src/main/java/school/faang/user_service/mapper/UserMapper.java:
```diff
@@ -1,10 +1,11 @@
 import school.faang.user_service.entity.User;
 
-@Mapper(componentModel = "spring")
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
```
рекомендую все таки привыкать делать через ReportingPolicy.ERROR иначе можно что нить упустить при мапенге когда сервис будет дальше развиваться

alobraztsov86 replied:
Поменял на ReportingPolicy.ERROR

Nafanuil commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -17,6 +18,13 @@ public interface GoalRepository extends JpaRepository<Goal, Long> {
+            JOIN user_goal ug ON g.id = ug.goal_id
+            WHERE ug.user_id = ?1
+            """)
+    List<Goal> findGoalsListByUserId(long userId);
```
не вижу что бы код использовался

alobraztsov86 replied:
удалил

Nafanuil commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,46 @@
+public class MentorshipService {
+
+    @Autowired
+    private UserRepository userRepository;
```
не используется

alobraztsov86 replied:
удалил

Nafanuil commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,46 @@
+     *
+     * @param userToDeactivateId - mentor who is being deactivated
+     */
+    public void stopMentorship(Long userToDeactivateId) {
```
а я разве могу захотеть  остановить менторство только по причине деактивации?

alobraztsov86 replied:
Я просто сделал, как задании написано(может ещё как-то сделать можно):
_После деактивации профиля: если пользователь был ментором других пользователей — необходимо остановить менторство. Для этого нужно написать соответствующий метод в MentorshipService. Ментор должен пропасть из списка менторов своих менти._

Nafanuil replied:
вопрос исключительно к названиям, ты же не пишешь метод который запустят 1 раз, его могут использовать и другие разработчики в дальнейшем, да может и ты сам его будешь использовать но уже не при деактивации пользователя

Nafanuil commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,109 @@
+        }
+    }
+
+    private List<Goal> findGoalsToDelete(Long userId, List<Goal> userGoals) {
```
зачем тебе параметр userId ?

Nafanuil replied:
что то не так либо с неймингом либо с методом
ты же просто отфильтровываешь цели по признаку - у цели есть 1 любой юзер

alobraztsov86 replied:
поменял одно условие(вернул как раньше было), чтобы однозначно читался код и название метода
вернул условие 
` if ((users.size() == 1) && (users.get(0).getId().equals(userId)) )`

Nafanuil commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,109 @@
+    private static final String ERROR_USER_IS_NOT_PRESENTED_IN_DB = "User is not presented in DB";
+    private static final String LOG_MESSAGE_DELETE_GOAL = "Deleting goals fot  userId={} goal={} ";
+
+    public void deactivateUser(Long userId) {
```
а где код который делает - При этом все цели, которые ментор поставил менти сохраняются, но больше не хранят mentorId — теперь они выглядят так, как будто менти поставил их себе сам.

alobraztsov86 replied:
Это было в методе stopMentorship() c 49 по 57 строку
Сейчас немного поменял код и названия переменных.

Nafanuil commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,109 @@
+    private static final String ERROR_USER_IS_NOT_PRESENTED_IN_DB = "User is not presented in DB";
+    private static final String LOG_MESSAGE_DELETE_GOAL = "Deleting goals fot  userId={} goal={} ";
+
+    public void deactivateUser(Long userId) {
```
тут есть еще одна проблемка которую вы пока не проходили, все это должно либо выполнится целиком либо вернутся к состоянию на начало метода, и это проще всего добиться делая все в 1 транзакции, можно либо поставить аннотацию и сделать ряд правок в коде, а можно попробовать работать с объектом userToDeactivate тк по факту в нем уде есть все те данные которые тебе нужно найти и поправить вместо того что бы работать с отдельными репозиториями

alobraztsov86 replied:
Беру на заметку, пока не знаю что с этим делать

Nafanuil commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -47,4 +55,6 @@ WITH RECURSIVE subtasks AS (
             """)
     List<User> findUsersByGoalId(long goalId);
+
+    List<Goal> findGoalsByMentorId(Long id);
```
юзер же содержит список целей, и ты его как раз берешь из базы, зачем тебе еще руками за ними ходить?

alobraztsov86 replied:
У меня тоже была такая мысль, но глядя на структуру таблиц goal and user_goal  разделил этот функционал, предположив, что у задачи может не быть ментора.
В описании задачи по целям говорится,что пользователь(не обязательно ментор) может быть создателем цели. К тому же в задаче по целям не прописано как к ним относится ментор.
https://faang-school.atlassian.net/browse/BJS2-65910


Если у любой задачи должен быть ментор, тогда перепишу код, и вытащу данные через **userToDeactivate** без доп. запроса в базу.

Nafanuil replied:
вот у тебя есть тут поле
![image](https://github.com/user-attachments/assets/81c05348-0fbd-4e5f-985b-f4efee6b6315)
 хоть оно и названо очень коряво но его все равно можно использовать чтобы получить цели у который пользователь ментор

Nafanuil commented on src/main/java/school/faang/user_service/repository/mentorship/MentorshipRepository.java:
```diff
@@ -1,7 +1,15 @@
+            WHERE mentor_id = :mentorId
+            """)
+    @Modifying
+    void deleteByMentorId( long mentorId);
```
можно конечно и так, но тут даже сам репозиторий намекает что с этой связью надо работать через юзера ( CrudRepository<User, Long> )

alobraztsov86 replied:
Похоже, удалил уже этот метод

Nafanuil replied:
я все еще вижу его

Commit:
BJS2-52697 Deactivate user. Review changes 2. MentorshipService.stopMentorship() updated; GoalRepository updated; UserController updated. Autowired deleted

Nafanuil commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,106 @@
+        log.info(LOG_MESSAGE_DEACTIVATING_STARTS, userId);
+        User userToDeactivate = getUserFromDataBase(userId);
+
+        quitGoals(userId, userToDeactivate.getGoals());
```
тк userToDeactivate содержит нужный id и список то если передать вместо двух параметров, только userToDeactivate смысл станет более понятный

alobraztsov86 replied:
поменял на userToDeactivate

alobraztsov86 replied:
Передал userToDeactivate

Nafanuil commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,106 @@
+                    .filter(u -> !u.getId().equals(userId))
+                    .collect(Collectors.toList());
+            goal.setUsers(usersOfGoalWithoutDeactivatedUser);
+            goalService.updateGoal(userId, goal);
```
а зачем там userId?

alobraztsov86 replied:
Думал, что для наглядности, чтобы видеть какой пользователь прекращает участие в цели

Nafanuil commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,106 @@
+        return getUserFromDataBase(userId);
+    }
+
+    private void quitGoals(Long userId, List<Goal> userGoals) {
```
если честно то я достаточно много времени потратил на то чтобы понять что именно тут делается, было бы гораздо проще если вынести это в два метода с говорящими названиями к примеру removeTasksWhereIamTheOnlyAssignee и removeMeFromTaskParticipants (привет chatGPT =) )
правда все равно останется не большой запашок, потому что если я поменяю местами эти два метода то код сломается

alobraztsov86 replied:
ChatGPT тут не использовался для написания кода)
- Разделил на два метода.
- В метод removeUserFromGoalParticipants()
добавил условие, чтобы удалялся пользователь из партисипантов цели, только если у цели более одного партисипанта.  тогда методы
removeGoalsWhereUserIsTheOnlyAssignee() и removeUserFromGoalParticipants()  можно менять местами и код не сломается.

Commit:
BJS2-52697 Deactivate user. Review changes 3. GoalSer
.deleteGoal() and EventSer
.deleteEvent() methods are implemented and code changed to work with real implementation.

Commit:
BJS2-52697 Deactivate user. Review changes 3. unused import deleted

alobraztsov86 commented:
Удаление Целей, смена ментора у цели
![image](https://github.com/user-attachments/assets/10762830-8a66-4bd8-b16e-d8950180a9a9)

Удаление Evens
![image](https://github.com/user-attachments/assets/ce4aabe2-5bdc-431e-91cd-31e06d63664c)

Методы 
```
 removeGoalsWhereUserIsTheOnlyAssigneeOrHasNoAssignee(user);
 removeUserFromGoalParticipants(user);
```
 менял местами в  `private void quitGoals(User user)` - работает

Commit:
BJS2-52697 Deactivate user. Review changes 3. unused import deleted 2
